### PR TITLE
SAMZA-2458: Update ProcessJobFactory and ThreadJobFactory to load full job config

### PR DESCRIPTION
**Pull request by kw2542**
Design:
https://cwiki.apache.org/confluence/display/SAMZA/SEP-23%3A+Simplify+Job+Runner

Changes:
1. Update ProcessJobFactory to load full job config, execute planning and write full job config back to coordiantor stream, which was done by RemoteApplicationRunner
2. Update ThreadJobFactory to load full job config, execute planning and write full job config back to coordiantor stream, which was done by RemoteApplicationRunner

These are needed as RemoteApplicationRunner no longer get full job config, no longer execute planning and no longer persist full job config in coordinator stream anymore.

API Changes:
None.

Upgrade Instructions:
See https://github.com/apache/samza/pull/1256 for more detailed instructions.

Usage Instructions:
See https://github.com/apache/samza/pull/1256 for more detailed instructions.

Tests
1. Local deploy a yarn job and verified it is working properly.

### CODE REVIEW

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 1]
```diff
@@ -51,15 +70,11 @@ class ProcessJobFactory extends StreamJobFactory with Logging {
     val coordinatorStreamStore: CoordinatorStreamStore = new CoordinatorStreamStore(config, new MetricsRegistryMap())
     coordinatorStreamStore.init()
 
-    val configFromCoordinatorStream: Config = CoordinatorStreamUtil.readConfigFromCoordinatorStream(coordinatorStreamStore)
-
     val changelogStreamManager = new ChangelogStreamManager(new NamespaceAwareCoordinatorStreamStore(coordinatorStreamStore, SetChangelogMapping.TYPE))
-
-    val coordinator = JobModelManager(configFromCoordinatorStream, changelogStreamManager.readPartitionMapping(),
-      coordinatorStreamStore, metricsRegistry)
+    val coordinator = JobModelManager(config, changelogStreamManager.readPartitionMapping(), coordinatorStreamStore, metricsRegistry)
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Let's just call this jobModelManager

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 2]
```diff
@@ -22,25 +22,44 @@ package org.apache.samza.job.local
...
 import org.apache.samza.metrics.MetricsRegistryMap
 import org.apache.samza.startpoint.StartpointManager
 import org.apache.samza.storage.ChangelogStreamManager
-import org.apache.samza.util.{CoordinatorStreamUtil, Logging, ReflectionUtil}
+import org.apache.samza.util.{ConfigUtil, CoordinatorStreamUtil, DiagnosticsUtil, Logging, ReflectionUtil}
 
 import scala.collection.JavaConversions._
 
 /**
  * Creates a stand alone ProcessJob with the specified config.
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Minor: Let's remove "stand alone"

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 3]
```diff
@@ -22,25 +22,44 @@ package org.apache.samza.job.local
...
  */
 class ProcessJobFactory extends StreamJobFactory with Logging {
-  def getJob(config: Config): StreamJob = {
+  def getJob(submissionConfig: Config): StreamJob = {
+    val originalConfig = ConfigUtil.loadConfig(submissionConfig)
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: It'll be useful to explain in the PR description why this change is necessary (under Changes), and what has changed from a user perspective (in the API changes section).

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 4]
```diff
@@ -22,25 +22,44 @@ package org.apache.samza.job.local
...
+
+    // Execute planning
+    val planner = new RemoteJobPlanner(ApplicationDescriptorUtil.getAppDescriptor(ApplicationUtil.fromConfig(originalConfig), originalConfig))
+    val jobConfigs = planner.prepareJobs
+
+    if (jobConfigs.size != 1) {
+      throw new SamzaException("Only single process job is supported.")
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: "Only single stage jobs are supported with ProcessJobFactory."

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala (hunk 1 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: updated, nice catch, not sure why it was named coordinator before.

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala (hunk 3 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Updated.

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 5]
```diff
@@ -22,25 +22,48 @@ package org.apache.samza.job.local
...
 class ProcessJobFactory extends StreamJobFactory with Logging {
-  def getJob(config: Config): StreamJob = {
+  def getJob(submissionConfig: Config): StreamJob = {
+    var config = submissionConfig
+
+    if (new JobConfig(submissionConfig).getConfigLoaderFactory.isPresent) {
+      val originalConfig = ConfigUtil.loadConfig(submissionConfig)
+
+      // Execute planning
+      val planner = new RemoteJobPlanner(ApplicationDescriptorUtil.getAppDescriptor(ApplicationUtil.fromConfig(originalConfig), originalConfig))
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: From the naming of `RemoteJobPlanner`, this seems out of place. I believe `RemoteJobPlanner` was intended for use in `RemoteApplicationRunner`, but theoretically, any `ApplicationRunner` can be used with any `StreamJobFactory`.
I guess the layers of abstraction are changing with the simplification of the job runners, but maybe we should rename some classes/interfaces and make sure the logic is in the right place as well.

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 6]
```diff
@@ -22,25 +22,48 @@ package org.apache.samza.job.local
...
+
+      if (jobConfigs.size != 1) {
+        throw new SamzaException("Only single process job is supported.")
+      }
+
+      // This is the full job config
+      config = jobConfigs.get(0)
+      // This needs to be consistent with RemoteApplicationRunner#run where JobRunner#submit to be called instead of JobRunner#run
+      CoordinatorStreamUtil.writeConfigToCoordinatorStream(config)
+      DiagnosticsUtil.createDiagnosticsStream(config)
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Similar to above: does this only need to be consistent with `RemoteApplicationRunner`?

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala [hunk 7]
```diff
@@ -87,13 +106,13 @@ class ProcessJobFactory extends StreamJobFactory with Logging {
     val commandBuilder = ReflectionUtil.getObj(commandBuilderClass, classOf[CommandBuilder])
 
     // JobCoordinator is stopped by ProcessJob when it exits
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Could you please update this comment? It looks like it was out-of-date before you made your changes, but would be nice to clarify it now.

File: samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala [hunk 8]
```diff
@@ -45,20 +47,36 @@ import scala.collection.mutable
...
 class ThreadJobFactory extends StreamJobFactory with Logging {
-  def getJob(config: Config): StreamJob = {
+  def getJob(submissionConfig: Config): StreamJob = {
     info("Creating a ThreadJob, which is only meant for debugging.")
+    var config = submissionConfig
+    if (new JobConfig(submissionConfig).getConfigLoaderFactory.isPresent) {
+      val originalConfig = ConfigUtil.loadConfig(submissionConfig)
+
+      // Execute planning
+      val planner = new RemoteJobPlanner(ApplicationDescriptorUtil.getAppDescriptor(ApplicationUtil.fromConfig(originalConfig), originalConfig))
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala]: Similar to above about `RemoteJobPlanner`

File: samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala [hunk 9]
```diff
@@ -45,20 +47,36 @@ import scala.collection.mutable
   * Creates a new Thread job with the given config
   */
 class ThreadJobFactory extends StreamJobFactory with Logging {
-  def getJob(config: Config): StreamJob = {
+  def getJob(submissionConfig: Config): StreamJob = {
     info("Creating a ThreadJob, which is only meant for debugging.")
+    var config = submissionConfig
+    if (new JobConfig(submissionConfig).getConfigLoaderFactory.isPresent) {
```
Comment (cameronlee314) [inline samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala]: This block seems duplicated with `ProcessJobFactory`. Any benefit in sharing code?

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala (hunk 5 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Agree, it is interesting that StreamFactory is only used in JobRunner, and JobRunner is only used in RemoteApplicationRunner, looks like we have excessive layers in our structure which should be simplified to avoid future confusion.

File: samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala (hunk 6 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/job/local/ProcessJobFactory.scala]: Yes, StreamFactory is only used by JobRunner, which is only used by RemoteApplicationRunner.

File: samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala (hunk 9 above)
Comment (kw2542) [reply samza-core/src/main/scala/org/apache/samza/job/local/ThreadJobFactory.scala]: I am planning to simplify and refactor all of them after I introduces beam related planning.
